src/components/Dropdown/Dropdown.tsx

```diff
@@ -0,0 +1,362 @@
+import { ComponentSlotStyle } from '../../themes/types'
+import Downshift, { DownshiftState, StateChangeOptions } from 'downshift'
+import Label from '../Label/Label'
+import { UIComponent, customPropTypes } from '../../lib'
+import Input from '../Input/Input'
+import keyboardKey from 'keyboard-key'
+import List from '../List/List'
+import Text from '../Text/Text'
+import Image from '../Image/Image'
+import ListItem from '../List/ListItem'
+import _ from 'lodash'
+import Icon from '../Icon/Icon'
+
+export interface DropdownProps {
+  className?: string
+  fluid?: boolean
+  items?: DropdownListItem[]
+  multiple?: boolean
+  onChange?: (active: DropdownListItem[]) => any
+  placeholder?: string
+  search?: boolean
+  styles?: ComponentSlotStyle<DropdownProps, DropdownState>
+}
+
+export interface DropdownState {
+  active: DropdownListItem[]
+  backspaceDelete: boolean
+  focused: boolean
+  inputValue?: string
+  message?: string
+}
+
+export interface DropdownListItem {
+  key: string
+  header: string
+  content?: string
+  image?: string
+}
+
+/**
+ */
+export default class Dropdown extends UIComponent<Extendable<DropdownProps>, DropdownState> {
+  private inputRef: HTMLElement
+
+  static displayName = 'Dropdown'
+
+  static className = 'ui-dropdown'
+
+  static propTypes = {
+    /** Additional CSS class name(s) to apply.  */
+    className: PropTypes.string,
+
+    /** An input can take the width of its container. */
+    fluid: PropTypes.bool,
+
+    /** Shorthand array of props for ListItem. */
+    items: customPropTypes.collectionShorthand,
+
+    /** A dropdown can have a multiple selection. */
+    multiple: PropTypes.bool,
+
+    /**
+     * Event for request to change 'open' value.
+     * @param {SyntheticEvent} event - React's original SyntheticEvent.
+     * @param {object} data - All props and proposed value.
+     */
+    onChange: PropTypes.func,
+
+    /** A message to serve as placeholder. */
+    placeholder: PropTypes.string,
+
+    /** A dropdown can have a search field. */
+    search: PropTypes.bool,
+
+    /** Additional CSS styles to apply to the component instance.  */
+    styles: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
+  }
+
+  state: DropdownState = {
+    active: this.props.multiple ? [] : null,
+    focused: false,
+    inputValue: '',
+    backspaceDelete: true,
+  }
+
+  public renderComponent({ ElementType, styles, variables }): React.ReactNode {
+    const { search, multiple } = this.props
+    const { inputValue } = this.state
+    // we hold active elemts in the array, downshift should not know anything.
+    const selectedPropIfMultiple = { ...(multiple && { selectedItem: null }) }
+
+    return (
+      <ElementType>
+        <Downshift
+          onChange={this.onDropdownChange}
+          inputValue={inputValue}
+          {...selectedPropIfMultiple}
+          stateReducer={this.stateReducer}
+          itemToString={(item: DropdownListItem) => (item ? item.header : '')}
```

Comment:
If we let consumers to pass `items` we should probably also allow them to pass `itemToString` (and maybe fallback to default if not provided). My worry is that if `item` is shorthand, it might be a string, component or functional component as well. @mnajdova we will need your advice here.

Reply:
added as optional prop